Feature/customer crud

michaelr1300 opened the pull request:
- MIGRATION table customer
- /customer
- Edit, Add, Delete using modal
- Routes for customer pagination

Sample data:
![image](https://user-images.githubusercontent.com/46708593/113562251-fdc35b80-962f-11eb-9698-c8f25ecf8129.png)

michaelr1300 committed:
Customer CRUD

- MIGRATION table customer
- /customer
- Edit, Add, Delete using modal
- Routes for customer pagination

michaelr1300 committed:
Rename Migration

bagaskarala commented:
@michaelr1300 menu sidebarnya kok gak ada

michaelr1300 commented:
Sidebar nya sekalian ada menu untuk invoice gapapa mas?

bagaskarala commented:
> Sidebar nya sekalian ada menu untuk invoice gapapa mas?

boleh boleh

-----

@michaelr1300 ini typenya sebaiknya pake lowercase
![image](https://user-images.githubusercontent.com/14112952/113700189-beb10b00-9700-11eb-9e35-7dc0915867fd.png)

bagaskarala commented:
@michaelr1300 
oiya kalo pattern-nya sebisa mungkin disamain yang lain gimana?

misal seperti ini untuk method `add`:
https://github.com/bagaskarala/sigap/blob/12d0fee867479cfbf9be079cb1b84913f61811f1/application/modules/library/controllers/Library.php#L21

misal untuk deklarasi rules validasi di model, ga perlu set rules satu, contohnya begini:
https://github.com/bagaskarala/sigap/blob/12d0fee867479cfbf9be079cb1b84913f61811f1/application/modules/library/models/Library_model.php#L5

tambahan:
kalau form pake modal lalu datanya gak valid, jadinya reload dan hilang semua perubahannya

bagaskarala commented:
kalo pake modal jadi kurang bagus jika ada error validasi. kalaupun mau pake modal, harusnya handlenya via ajax semua, biar gak reload dan hilang.
https://user-images.githubusercontent.com/14112952/113703765-79430c80-9705-11eb-9676-0449d9d74148.mp4

michaelr1300 committed:
Update Customer CRUD

- Add sidebar
- Change data to use lowercase
- Move validation do model
- Rename functions in controller
- Add Ajax validation
- Rename HTML components

michaelr1300 commented on application/modules/customer/controllers/Customer.php:
```diff
@@ -17,14 +17,14 @@ public function index($page = null)
+            'reseller'      => 'Reseller',
+            'penulis'        => 'Penulis',
+            'member'        => 'Member',
+            'biasa'        => ' - '
```
type pakai lowercase

michaelr1300 commented on application/modules/customer/models/Customer_model.php:
```diff
@@ -5,46 +5,66 @@ class Customer_model extends MY_Model
-        $this->db->where('customer_id', $id);
-        $this->db->update('customer');
-        return TRUE;
+        $data = array();
```
Validation di modal untuk AJAX

michaelr1300 commented on application/modules/customer/models/Customer_model.php:
```diff
@@ -5,46 +5,66 @@ class Customer_model extends MY_Model
+        {
+            echo json_encode($data);
+            exit();
+        }
```
Validation AJAX untuk edit (karena id input berbeda dengan modal Add)

michaelr1300 commented on application/modules/customer/views/index_customer.php:
```diff
@@ -212,6 +212,7 @@ class="font-weight-bold"
                                             id="name"
                                             class="form-control"
                                         />
+                                        <small id="error-name" class="d-none error-message text-danger">Nama wajib diisi!</small>
```
Error message yang akan ditampilkan saat validation error

michaelr1300 commented on application/modules/customer/models/Customer_model.php:
```diff
@@ -5,46 +5,66 @@ class Customer_model extends MY_Model
+        {
+            echo json_encode($data);
+            exit();
+        }
```
Validation rule untuk modal add (AJAX)

michaelr1300 commented on application/modules/customer/views/index_customer.php:
```diff
@@ -239,6 +240,7 @@ class="font-weight-bold"
                                             id="phone-number"
                                             class="form-control"
                                         />
+                                        <small id="error-phone-number" class="d-none error-message text-danger">Nomor telepon wajib diisi!</small>
```
Error message yang akan ditampilkan saat validation error

michaelr1300 commented on application/modules/customer/models/Customer_model.php:
```diff
@@ -5,46 +5,66 @@ class Customer_model extends MY_Model
-        $this->db->set($update);
-        $this->db->where('customer_id', $id);
-        $this->db->update('customer');
-        return TRUE;
```
Function dipindah langsung di controller supaya se-pattern dengan fitur yang lain

michaelr1300 commented on application/modules/customer/views/index_customer.php:
```diff
@@ -381,16 +387,83 @@ function modalEdit(customerId) {
+            alert(err)
+        },
+    });
+})
```
Sumbit modal add
contoller akan run validation
- Jika ada validation error, controller return array berisi id yang error. Jquery mencari id yang error dari array, kemudain remove class `d-none` sehingga error message muncul di modal. Saat user submit ulang, error message di hidden kemudian controller cek ulang validation, jika masih ada error maka ditampilkan lagi error message nya
- Jika tidak ada validation error, refresh page

https://user-images.githubusercontent.com/46708593/113804311-835a1f00-9788-11eb-82fb-ca862d38f5eb.mp4

michaelr1300 commented on application/modules/customer/views/index_customer.php:
```diff
@@ -381,16 +387,83 @@ function modalEdit(customerId) {
+            alert(err)
+        },
+    });
+})
```
Submit modal edit, sistemnya sama seperti submit modal add

michaelr1300 commented on application/modules/customer/controllers/Customer.php:
```diff
@@ -37,60 +37,52 @@ public function index($page = null)
-        $this->load->library('form_validation');
-        $this->form_validation->set_rules('name', 'Nama Customer', 'required');
-        $this->form_validation->set_rules('type', 'Tipe Customer', 'required');
-        $this->form_validation->set_rules('phone-number', 'Nomor HP Customer', 'required');
```
Validation dipindah ke model

bagaskarala commented on application/modules/customer/models/Customer_model.php:
```diff
@@ -5,46 +5,66 @@ class Customer_model extends MY_Model
-        $this->db->update('customer');
-        return TRUE;
+        $data = array();
+        $data['inputerror'] = array();
```
ini dibuat lebih readable ya, misal `input_error` begitu

bagaskarala commented on application/modules/customer/models/Customer_model.php:
```diff
@@ -5,46 +5,66 @@ class Customer_model extends MY_Model
     public $per_page;
 
-    public function add_customer()
+    public function validateModalAdd()
```
untuk method di php sebaiknya pakai snake_case aja ya biar setara, misal `validate_modal_add`

bagaskarala commented on application/modules/customer/controllers/Customer.php:
```diff
@@ -17,14 +17,14 @@ public function index($page = null)
-            'Penulis'        => 'Penulis',
-            'Member'        => 'Member',
-            'Biasa'        => ' - '
+            'distributor'      => 'Distributor',
+            'reseller'      => 'Reseller',
+            'penulis'        => 'Penulis',
+            'member'        => 'Member',
+            'biasa'        => ' - '
```
sebisa mungkin pake english semua ya biar konsisten, `penulis` jadi `author`. 
Lalu `biasa` itu maksudnya apa kok pake `-` ? kalo mau default mungkin pake `default` atau `general`.

michaelr1300 replied:
Biasa itu untuk orang umum mas. Di controller diganti jadi `general`

michaelr1300 committed:
Fix Naming

- Rename function to snake case
- Rename customer type
- Rename validation variable to be more readable

bagaskarala commented on application/modules/customer/views/index_customer.php:
```diff
@@ -0,0 +1,469 @@
+$i = isset($page) ? $page * $per_page - $per_page : 0;
+
+
+$type_options = [
+    ''                  => '- Filter Jenis Customer -',
+    'distributor'       => 'Distributor',
+    'reseller'          => 'Reseller',
+    'author'            => 'Penulis',
+    'member'            => 'Member',
+    'general'           => 'Umum'
+];
```
@michaelr1300 ini daripada deklarasi ulang, lebih baik pake data dari controller, biar pakai satu sumber data.
```php
$type_options =  array_merge(['' => '- Filter Jenis Customer -'], $customer_type);
```

michaelr1300 replied:
Oke mas

michaelr1300 committed:
Update dropdown type option filter

- Use data from controller
